File: src/main/java/editor/gui/ccp/handler/CategoryImportHandler.java
package editor.gui.ccp.handler;

import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.io.IOException;
import java.util.function.Predicate;

import javax.swing.TransferHandler;

import editor.collection.deck.Category;
import editor.gui.ccp.data.CategoryTransferData;
import editor.gui.ccp.data.DataFlavors;

/**
 * Importer for category specifications.
 * 
 * @author Alec Roelke
 */
public class CategoryImportHandler extends TransferHandler implements ImportHandler
{
    /** Function specifying how to tell if the deck contains the category already. */
    private final Predicate<Category> contains;
    /** Function specifying how to add the category. */
    private final Predicate<Category> add;

    /**
     * Create a new category import handler.
     * 
     * @param c function specifying whether or not the source component contains the
     * category, if applicable (should just return false if it's not applicable)
     * @param a function specifying how to add the category to the component
     */
    public CategoryImportHandler(Predicate<Category> c, Predicate<Category> a)
    {
        contains = c;
        add = a;
    }

    @Override
    public DataFlavor supportedFlavor()
    {
        return DataFlavors.categoryFlavor;
    }

    @Override
    public boolean canImport(TransferHandler.TransferSupport supp)
    {
        if (supp.isDrop())
            return false;
        else if (supp.isDataFlavorSupported(DataFlavors.categoryFlavor))
        {
            try
            {
                CategoryTransferData data = (CategoryTransferData)supp.getTransferable().getTransferData(DataFlavors.categoryFlavor);
                return !contains.test(data.spec);
            }
            catch (UnsupportedFlavorException | IOException e)
            {
                return false;
            }
        }
        else
            return false;
    }

    @Override
    public boolean importData(TransferHandler.TransferSupport supp)
    {
        try
        {
            if (!canImport(supp))
                return false;
            else
                return add.test(((CategoryTransferData)supp.getTransferable().getTransferData(DataFlavors.categoryFlavor)).spec);
        }
        catch (UnsupportedFlavorException | IOException e)
        {
            return false;
        }
    }
}

File: src/main/scala/editor/gui/ccp/handler/CategoryImportHandler.scala
package editor.gui.ccp.handler

import editor.collection.Categorization
import editor.gui.ccp.data.CategoryTransferData
import editor.gui.ccp.data.DataFlavors

import java.awt.datatransfer.UnsupportedFlavorException
import java.io.IOException
import javax.swing.TransferHandler

/**
 * Import handler for [[Categorization]]s.  Used only for cut/copy/paste; does not support drag-and-drop.
 * 
 * @constructor create a new category import handler that processes categories with arbitrary predicates
 * @param contains predicate indicating if the category is already contained by the target element; duplicate categories
 * are not allowed
 * @param add function for adding the category to the element and returning true on success and false otherwise
 * 
 * @author Alec Roelke
 */
class CategoryImportHandler(contains: (Categorization) => Boolean, add: (Categorization) => Boolean) extends TransferHandler with ImportHandler {
  @deprecated def this(contains: java.util.function.Predicate[Categorization], add: java.util.function.Predicate[Categorization]) = this(contains.test, add.test)

  override def supportedFlavor = DataFlavors.categoryFlavor

  override def canImport(supp: TransferHandler.TransferSupport) = if (supp.isDrop) false else if (supp.isDataFlavorSupported(supportedFlavor)) {
    try {
      val data = supp.getTransferable.getTransferData(supportedFlavor) match {
        case c: CategoryTransferData => c
        case x => throw ClassCastException(s"expected ${classOf[CategoryTransferData]}, got ${x.getClass}")
      }
      !contains(data.spec)
    } catch {
      case _: UnsupportedFlavorException => false
      case _: IOException => false
    }
  } else false

  override def importData(supp: TransferHandler.TransferSupport) = if (!canImport(supp)) false else {
    try {
      val data = supp.getTransferable.getTransferData(supportedFlavor) match {
        case c: CategoryTransferData => c
        case x => throw ClassCastException(s"expected ${classOf[CategoryTransferData]}, got ${x.getClass}")
      }
      add(data.spec)
    } catch {
      case _: UnsupportedFlavorException => false
      case _: IOException => false
    }
  }
}